File: algorithm_library/python/graph/LCA.py
from collections import deque

class LCA:
    def __init__(self, N, G, root):  
        # 根からの距離と親を求める
        self.N = N
        self.bit = N.bit_length() + 2
        self.parent = [[-1]*(N+1) for _ in range(self.bit)]
        self.parent[0][root] = -1
        self.depth = [-1]*(N+1) # 根からの距離
        self.depth[root] = 0
        que = deque()
        que.append(root)
        while que:
            v = que.popleft()
            for v2 in G[v]:
                if self.depth[v2]==-1:
                    self.depth[v2] = self.depth[v]+1
                    self.parent[0][v2] = v
                    que.append(v2)

        # ダブリングする
        for i in range(1,self.bit):
            for j in range(N+1):
                if self.parent[i-1][j]!=-1: 
                    self.parent[i][j] = self.parent[i-1][self.parent[i-1][j]]

    # u と v の LCA を返す
    # O(log N)
    def lca(self,u,v):
        # 深さを同じにする
        if self.depth[u]<self.depth[v]: u,v = v,u
        for k in range(self.bit):
            if ((self.depth[u]-self.depth[v]) >> k) & 1:
                u = self.parent[k][u]
        # 二分探索する
        if u==v: return u
        for k in reversed(range(self.bit)):
            if self.parent[k][u] != self.parent[k][v]:
                u = self.parent[k][u]
                v = self.parent[k][v]
        return self.parent[0][u]

    # u と v の距離を返す
    def get_distance(self,u,v):
        return self.depth[u] + self.depth[v] - 2*self.depth[self.lca(u,v)]
    
    # u-v のパス上に a が存在するか
    def is_on_path(self,u,v,a):
        return self.get_distance(u,a) + self.get_distance(v,u) == self.get_distance(u,v)

File: algorithm_library/cpp-from-py/graph/LCA.cpp
#include <bits/stdc++.h>
using namespace std;

struct LCA {
    int N, bit;
    vector<vector<int>> parent;
    vector<int> depth;

    LCA(int N_, const vector<vector<int>>& G, int root) : N(N_) {
        bit = 32 - __builtin_clz(N) + 2;
        parent.assign(bit, vector<int>(N + 1, -1));
        depth.assign(N + 1, -1);
        depth[root] = 0;
        queue<int> que;
        que.push(root);
        while (!que.empty()) {
            int v = que.front();
            que.pop();
            for (int v2 : G[v]) {
                if (depth[v2] == -1) {
                    depth[v2] = depth[v] + 1;
                    parent[0][v2] = v;
                    que.push(v2);
                }
            }
        }
        for (int i = 1; i < bit; i++) {
            for (int j = 0; j <= N; j++) {
                if (parent[i - 1][j] != -1) parent[i][j] = parent[i - 1][parent[i - 1][j]];
            }
        }
    }

    int lca(int u, int v) const {
        if (depth[u] < depth[v]) swap(u, v);
        for (int k = 0; k < bit; k++) if (((depth[u] - depth[v]) >> k) & 1) u = parent[k][u];
        if (u == v) return u;
        for (int k = bit - 1; k >= 0; k--) {
            if (parent[k][u] != parent[k][v]) {
                u = parent[k][u];
                v = parent[k][v];
            }
        }
        return parent[0][u];
    }

    int get_distance(int u, int v) const {
        return depth[u] + depth[v] - 2 * depth[lca(u, v)];
    }

    bool is_on_path(int u, int v, int a) const {
        return get_distance(u, a) + get_distance(a, v) == get_distance(u, v);
    }
};
